### ISSUE TITLE
cron module breaks with busybox 'crontab'

### ISSUE THREAD

**Issue by agaffney** [ansible/ansible-modules-core#2807]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

cron module
##### ANSIBLE VERSION

N/A
##### SUMMARY

When using the `cron` module on a target with busybox `crond`, the module fails to update the crontab. The error is cryptic, since all you see is the usage message from `crontab`.

After a bit of poking, it appears that the `cron` module wants to run the command `crontab -u someuser /path/to/new/crontab`, but the busybox `crontab` command only accepts options _after_ the path to the new crontab file on the commandline (i.e. `crontab /path/to/new/crontab -u someuser`).

**Comment by tyleha** [comment]
Fixed by https://github.com/ansible/ansible-modules-core/pull/2768, at least for me.

**Comment by AMDmi3** [comment]
This change breaks at least FreeBSD and was reverted in b5e012b.

As far as I can see, even busybox docs suggest that file should come after user argument: https://git.busybox.net/busybox/tree/miscutils/crontab.c#n14

so this change was bogus in the first place, or it was already fixed in busybox.

Update: checked with busybox 1.24.2: crontab works correctly with common argument order (`crontab -u someuser /path/to/new/crontab`)

**Comment by agaffney** [comment]
The version of busybox that I ran into this issue on was 1.23.2 on OpenWrt (15.05 Chaos Calmer). I know that @bcoca tested the new argument order in a few places, and I'm pretty sure both Linux and BSD of some sort were included. However, there are many different cron daemons, so I'm not surprised that there is some combination on which the original fix didn't work properly.

**Comment by AMDmi3** [comment]
I'm surprised it was needed somewhere in the first place. Order where flag arguments come before free arguments is common and fundamental, not supporting it and requiring reverse order is definitely a bug.

**Comment by agaffney** [comment]
I don't disagree that it's a bug in busybox. However, I can't easily upgrade the version of busybox on my router without going to a whole new OpenWrt release, which would then involve setting up some "extra" stuff again from scratch (perhaps with ansible, but that's not ready right now), so I'm "stuck".

It's unclear whether this was a recently-introduced bug in busybox or a long-standing one, but it does appear that it has been fixed in newer versions.

**Comment by AMDmi3** [comment]
The diff for crontab.c between busybox 1.23.2 and 1.24.2 is a single empty line removed, so I don't think that 1.23.2 crontab behavior differs from 1.24.2 in any way. See @tylerhartley comment as well - my guess is that the problem was never in the argument order.

**Comment by agaffney** [comment]
Yeah, something is really weird. I'm doing manual tests now on my router with `crontab -u nobody /path/to/file` and it seems to work fine. I _know_ that the original fix worked for me and that ansible was throwing errors prior to the fix, but I can't explain it now.

**Comment by agaffney** [comment]
Ah, I was wrong about which device and busybox version I encountered this on. This was actually on my QNAP NAS using busybox 1.01.

```
[~] # crontab --help
BusyBox v1.01 (2015.09.09-17:23+0000) multi-call binary

Usage: crontab [-c dir] {file|-}|[-u|-l|-e|-d user]

        file <opts>  replace crontab from file
        -    <opts>  replace crontab from stdin
        -u user      specify user
        -l [user]    list crontab for user
        -e [user]    edit crontab for user
        -d [user]    delete crontab for user
        -c dir       specify crontab directory
```

**Comment by AMDmi3** [comment]
God, busybox 1.01 was released in 2005.

**Comment by agaffney** [comment]
Tell that to QNAP :)

The firmware on this device was updated around a year ago, so they're still building their firmware with an ancient busybox, at least for older devices.

**Comment by bcoca** [comment]
this 'fix' was reverted as it broke FreeBSD cron (see #3750)

need new fix that detects cron type

**Comment by agaffney** [comment]
I was able to work around this by replacing the `/usr/bin/crontab -> /bin/busybox` symlink with this wrapper script (using ansible, of course):

```
#!/bin/sh

CRONTAB="busybox crontab"

CRON_ARGS_DIR=
CRON_ARGS_FILE=
CRON_ARGS_FLAGS=

while [ $# -ge 1 ]; do
        case $1 in
                -c)
                        CRON_ARGS_DIR="-c $2"
                        shift 2
                        ;;
                -[a-z])
                        CRON_ARGS_FLAGS="${CRON_ARGS_FLAGS} $1"
                        if [ -z "$2" ] || echo "$2" | grep -q '^-[a-z]'; then
                                shift
                        else
                                CRON_ARGS_FLAGS="${CRON_ARGS_FLAGS} $2"
                                shift 2
                        fi
                        ;;
                *)
                        CRON_ARGS_FILE="${CRON_ARGS_FILE} $1"
                        shift
                        ;;
        esac
done

eval $CRONTAB $CRON_ARGS_DIR $CRON_ARGS_FILE $CRON_ARGS_FLAGS
```

**Comment by agaffney** [comment]
I'm okay with closing this issue as "won't fix", since it only happens on an ancient busybox, I'm apparently the first to run into it, and I have a functional workaround to the problem.

**Comment by ansibot** [comment]
@dsummersl, @EvanK, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by EvanK** [comment]
Let's close this out, then? wontfix
